barretenberg/cpp/src/barretenberg/flavor/ultra.hpp

```diff
@@ -280,6 +283,103 @@ class UltraFlavor {
+            add_plookup_memory_records_to_wire_4(eta);
+        }
+
+        /**
+         * @brief Add plookup memory records to the fourth wire polynomial
+         *
+         * @details This operation must be performed after the first three wires have been committed to, hence the
+         * dependence on the `eta` challenge.
+         *
+         * @tparam Flavor
+         * @param eta challenge produced after commitment to first three wire polynomials
+         */
+        void add_plookup_memory_records_to_wire_4(FF eta)
+        {
+            // The plookup memory record values are computed at the indicated indices as
+            // w4 = w3 * eta^3 + w2 * eta^2 + w1 * eta + read_write_flag;
+            // (See plookup_auxiliary_widget.hpp for details)
+            auto wires = this->get_wires();
+
+            // Compute read record values
+            for (const auto& gate_idx : this->memory_read_records) {
+                wires[3][gate_idx] += wires[2][gate_idx];
+                wires[3][gate_idx] *= eta;
+                wires[3][gate_idx] += wires[1][gate_idx];
+                wires[3][gate_idx] *= eta;
+                wires[3][gate_idx] += wires[0][gate_idx];
+                wires[3][gate_idx] *= eta;
+            }
+
+            // Compute write record values
+            for (const auto& gate_idx : this->memory_write_records) {
+                wires[3][gate_idx] += wires[2][gate_idx];
+                wires[3][gate_idx] *= eta;
+                wires[3][gate_idx] += wires[1][gate_idx];
+                wires[3][gate_idx] *= eta;
+                wires[3][gate_idx] += wires[0][gate_idx];
+                wires[3][gate_idx] *= eta;
+                wires[3][gate_idx] += 1;
+            }
+        }
+
+        void compute_grand_product_polynomials(RelationParameters<FF>& relation_parameters)
+        {
+            auto public_input_delta = compute_public_input_delta<UltraFlavor>(this->public_inputs,
+                                                                              relation_parameters.beta,
+                                                                              relation_parameters.gamma,
+                                                                              this->circuit_size,
+                                                                              this->pub_inputs_offset);
+            relation_parameters.public_input_delta = public_input_delta;
+            auto lookup_grand_product_delta = compute_lookup_grand_product_delta(
+                relation_parameters.beta, relation_parameters.gamma, this->circuit_size);
+            relation_parameters.lookup_grand_product_delta = lookup_grand_product_delta;
+
+            // Compute permutation and lookup grand product polynomials
+            auto prover_polynomials = ProverPolynomials(this);
+            compute_grand_products<UltraFlavor>(*this, prover_polynomials, relation_parameters);
```

Comment:
with the exception of UltraFlavor here (and above in compute_public_input_delta ) instead of GoblinUltraFlavor